src/util/ecdsa.rs

```diff
@@ -412,6 +416,72 @@ impl<'de> ::serde::Deserialize<'de> for PublicKey {
     }
 }
 
+/// An ECDSA signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct EcdsaSig {
+    /// The underlying ECDSA Signature
+    pub sig: secp256k1::Signature,
+    /// The corresponding hash type
+    pub hash_ty: LegacySigHashType,
+}
+
+impl EcdsaSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, EcdsaSigError> {
+        match sl.split_last() {
+            Some((hash_ty, sig)) => {
+                let hash_ty = LegacySigHashType::from_u32_standard(*hash_ty as u32)
+                    .map_err(|_| EcdsaSigError::NonStandardSigHashType(*hash_ty))?;
+                let sig = secp256k1::Signature::from_der(sig)?;
+                Ok(EcdsaSig { sig, hash_ty })
+            },
+            None => {
+                Err(EcdsaSigError::Secp256k1(secp256k1::Error::InvalidSignature))?
```

Comment:
Well, it is not an ECDSA signature which is invalid, it is bitcoin-specific tx input serialization which misses sighashtype. So I think this should be a dedicated error type.

Reply:
I thought about this, and I think we can get away from this here. From an implementation point of view, it appears that we are missing the last byte from the slice (because the slice is empty), but that also means the underlying signature is invalid. 

There is really no way to know if we are missing the first byte of the sig, or the last byte (sighash byte)

Reply:
Agree that we can't detect is signature is wrong or sighashtype is absent; but it is clearly not a Secp error but a dedicated variant `InvalidSigOrMissedSigHash` :)

Reply:
I don't mean to be pedantic here. Would it be okay if I replace the following 
```diff
- Err(EcdsaSigError::Secp256k1(secp256k1::Error::InvalidSignature))?
+ secp256k1::Signature::from_slice(sl)?
```

I don't want to add unnecessary error variants that don't represent any meaning for the caller. I have partially started changing some code in `rust-miniscript` to actually deal with error types sanely rather than just propagating them.

Reply:
I am neither won't to be pedantic. I am just not follow your line of thoughts.

From what I see, there is a special data type representing the way ECDSA signature AND sighash are present in witness/scriptSig data - that is why we introduce these two new types.

And we can't distinguish situation from when the signature is badly serialized, contains invalid public keys or sighashtype is not present in ECDSA. These all are indistinguishable error types AND it is not the same as usual `secp256k1::Error` - none of its variants.

The second option for ECDSA witness/scriptSig data is that we have a proper signature and improper sighashtype (non-standard), i.e. existing `EcdsaSigError::NonStandardSigHashType`.

So I see having 
```rust
pub enum EcdsaSigError {
    NonStandardSigHashType,
    UnparsableData
}
```
to be more specific, not less specific.

With `SchnorrSig` we have different situation. There _we can_ distinguish improperly serialized signature from invalid `SigHashType`, since the signature always have 64 bytes in length. Thus, we have a different set of error variants, and there it will be correct to wrap `secp256k1::Error` for signature deserialization and add second option `InvalidSigHashType`. However, the parsing algorithm should be different from the one used for `EcdsaSig`.

Thus is a summary of all my comments in this PR.

Reply:
100% Agreed with this suggestion.